### PR TITLE
Improve precision of surface3d worldPosition calculation

### PR DESCRIPTION
**Pull request by hborchardt**
When the objectOffset is very large compared to the values in the localCoordinate, adding it to the localCoordinate loses precision in the localCoordinate. As the addition of the objectOffset was immediately undone by application of the model matrix, it is better to combine the model matrix and the translation by the objectOffset into one matrix and apply it directly to the localCoordinate.

Again, it's an issue mostly present when rendering data from a small time range.

Before: 
image

After:
image

### CODE REVIEW

**Comment by archmoj** [comment]
Awesome. I was thinking about this. Thank you very much!

**Comment by archmoj** [comment]
Please download and extract `baselines.tar` from https://app.circleci.com/pipelines/github/plotly/plotly.js/10479/workflows/4c357002-3abe-4fe6-8cda-e8f17a07c65f/jobs/229960/artifacts into `test/image/baselines`.
Thank you!

File: stackgl_modules/package.json [hunk 1]
```diff
@@ -28,7 +28,7 @@
     "gl-spikes2d": "^1.0.2",
     "gl-spikes3d": "^1.0.11",
     "gl-streamtube3d": "^1.4.2",
-    "gl-surface3d": "^1.6.1",
+    "gl-surface3d": "github:hborchardt/gl-surface3d#6c75a5b44781c2aaec543237312df1f14274feac",
```
Comment (archmoj) [inline stackgl_modules/package.json:31]: Please open a PR to `gl-vis/gl-surface3d` and provide the link here.
Thank you!

Comment (hborchardt) [reply stackgl_modules/package.json:31]: See https://github.com/gl-vis/gl-surface3d/pull/21

**Comment by archmoj** [comment]
Some tests are failing. 
Please download and replace new baselines from https://app.circleci.com/pipelines/github/plotly/plotly.js/10480/workflows/d5a9d293-c4d0-43aa-a838-3743b9609162/jobs/229984/artifacts this time.
Thank you!

**Comment by archmoj** [comment]
To lock this bug for happening again, I suggest you add a new mock inside `test/image/mock` forlder. Something similar to what posted into the PR description should be enough.
Please start the name of your new mock with `zz-gl3d_` (e.g. `zz-gl3d_surface_big-number.json``) so that it does not change the order of rendering of parallel runs on CircleCI.
After you pushed you could collect the baseline from the `arftifacts` tab of `test-baselines` and commit it inside `test/image/baselines` folder.
Thank you!

**Comment by hborchardt** [comment]
> Some tests are failing. Please download and replace new baselines from https://app.circleci.com/pipelines/github/plotly/plotly.js/10480/workflows/d5a9d293-c4d0-43aa-a838-3743b9609162/jobs/229984/artifacts this time. Thank you!

I accidentally reverted the change to the `stackgl_modules/index.js` file in the last commit. Now it should be fine.

File: stackgl_modules/package.json (hunk 1 above)
Comment (archmoj) [reply stackgl_modules/package.json:31]: `gl-surface3d@1.6.2` is published. Please install new patch.

**Comment by archmoj** [comment]
I'd love to include this PR in the upcoming `plotly.js` release.
@hborchardt Wondering if you are going to complete this PR today? 
I.e. install new `gl-surface3d` package and if possible adding a test.
If not, I may add those commits.
Thank you!

**Comment by hborchardt** [comment]
Hi, I can do it later today. When building the test case I saw that there
are other problems when the time range is on the order seconds or minutes,
and I was playing around trying if I can fix it as well, but it is more
complicated. So I will add a test with a time range of about an hour which
should be good.

Mojtaba Samimi ***@***.***> schrieb am Fr. 17. Mai 2024 um
17:46:

**Comment by archmoj** [comment]
Thanks for the quick reply.
Yes let's keep these PRs focused. 
So yeah this looks ready to merge when you install `gl-surface3d` patch.
But no need to rush it as I could simply wait for 3 hours. 
Thank you!

**Comment by hborchardt** [comment]
For reference, this is the added test image rendered with current `dist/plotly.js`
image

**Comment by archmoj** [comment]
Excellent bug fix!
Nicely done. 
:dancer:

**Comment by archmoj** [comment]
Plotly.js v2.33.0 is out: https://github.com/plotly/plotly.js/releases/tag/v2.33.0
Thanks for your contribution.

**Pull request merged**